### PR TITLE
Remote execution break mode

### LINKED ISSUE #21: Any way to run android gradle "install" commands locally?
One side effect of running `./gradlew install{flavor/buildtype/whatever}` is that the install task fails with a `DeviceException: No connected devices!` error (because the install task is running on the remote machine, which isn't running the emulator).

An obvious workaround is to just use an `assembe` task instead and then manually install/run the APK locally after the Gradle task is finished, but I'm curious if there's a workaround or if you guys have a better solution. Thanks!

### PR DESCRIPTION
**Pull request by skrugly**
#21 WIP

Feedback is welcome!

### Speed up Android build by breaking remote execution on `package` task 

The main result of Android application build is `.apk` file which contains all the code and resources of the application.
 `.apk` file is a zip archive that can't be downloaded from remote machine by rsync incrementally.
 Making just a small fix to codebase or adding new resource will lead to the need of downloading the whole file the size of which can be quite large.
 
The idea of breaking remote building process on `package` task is to stop remote build one step before archiving all the application files into `.apk` file and let rsync incrementally download only these files that changed during the build.
One line fix usually affects one `.dex` file whose size is significantly less than size of whole `.apk` file.
 
Mirakle will download all the inputs of `package` task and execute it by Gradle on local machine.        
      
```groovy
initscript {
    repositories {
        jcenter()
    }
    dependencies {
        classpath "com.instamotor:mirakle:1.4.0-RC-1"
    }
}
 
apply plugin: MirakleBreakMode // <- 
 
rootProject {
    mirakle {
        host "your_remote_machine"
        breakOnTasks += ["package"] // this is regex pattern that matches all build flavour variations of package task.  
        excludeRemote += ["build/tmp", "build/generated", "build/intermediates/*", "build/kotlin"]
        rsyncFromRemoteArgs += ["-i", "-c", "--compress-level=2"] 
    }
}
```

```
> ./gradlew assembleStagingDebug

:uploadToRemote
:executeOnRemote
Mirakle will break remote execution on task ':app:packageStagingDebug'
:downloadFromRemote
>fcst....... app/build/intermediates/dex/stagingDebug/mergeDexStagingDebug/classes2.dex
app:packageStagingDebug

Task uploadToRemote took : 0.19 secs
Task executeOnRemote took : 8.08 secs
Task downloadFromRemote took : 1.065 secs
Task app:packageStagingDebug took : 1.362 secs
```

### CODE REVIEW

File: plugin/src/main/kotlin/Mirakle.kt [hunk 1]
```diff
@@ -207,6 +305,13 @@ class Mirakle : Plugin<Gradle> {
     }
 }
 
+class MirakleBreakMode : Mirakle() {
+    override fun apply(gradle: Gradle) {
+        gradle.startParameter.projectProperties = gradle.startParameter.projectProperties.plus(BREAK_MODE to "true")
```
Comment (JayNewstrom) [inline plugin/src/main/kotlin/Mirakle.kt]: Could this just be done in the main plugin based on either a new flag, or `breakOnTasks` being non empty?

**Review by JayNewstrom — COMMENTED** [review]
The idea of this is really cool!

File: plugin/src/main/kotlin/Mirakle.kt (hunk 1 above)
Comment (skrugly) [reply plugin/src/main/kotlin/Mirakle.kt]: No. The problem is that mirake needs to know whether to start or not in break mode before project evaluation and before that moment when `mirakle {}` extension will be available. 
If build launched in break mode, mirakle will let Gradle populate task graph and thus spend some time on local machine.
Configuring mirakle as soon as possible through `gradle.startParameter.projectProperties` will save that time for non-break builds.

Comment (JayNewstrom) [reply plugin/src/main/kotlin/Mirakle.kt]: Makes sense. Thanks for the explanation.

Comment (skrugly) [reply plugin/src/main/kotlin/Mirakle.kt]: Maybe there is a better way to configure it, but I don't see it. Will be glad to hear your ideas.

Comment (anoopc1) [reply plugin/src/main/kotlin/Mirakle.kt]: If I understand it correctly,

A good side effect of current implementation is that it should be possible to launch in BreakMode vs Non-BreakMode without modifying source code by using `apply plugin: Mirakle` and passing `-Pmirakle.break.mode=true/false` while invoking the gradle-build from command-line.

Comment (skrugly) [reply plugin/src/main/kotlin/Mirakle.kt]: Yup, and I realized that parameter value from CLI should have priority over MirakleBreakMode. Just made a fix for this.

**Comment by dsvoronin** [comment]
`Task execution graph contains more than 1 task to break on. That is not supported by Mirakle yet.`
It would be handy to print these tasks right in the error context

**Comment by skrugly** [comment]
All right, for now you can find these tasks running build with `--dry-run`

**Comment by marcorighini** [comment]
@Adambl Great idea! Is this PR supposed to be merged soon?

**Comment by skrugly** [comment]
@marcorighini I'm looking for time to finish up the remaining tasks on this PR and will merge after that.
So far you can enjoy using Release Candidate version `1.4.0-RC-1`.

**Pull request merged**